### PR TITLE
libavfilter: add sobel filter optimazation with intel AVX512 VNNI

### PR DESCRIPTION
**Pull request by bwang30**
Sobel filter optimazation with intel AVX512 VNNI, performance data is here:
[<html xmlns:m="http://schemas.microsoft.com/office/2004/12/omml"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=PowerPoint.Slide>
<meta name=Generator content="Microsoft PowerPoint 15">
<style>

</style>
</head>

<body>

sobel | ffmpeg   filter 1T | ffmpeg filter xT | AVX512+VNNI   1T | AVX512+VNNI   xT
-- | -- | -- | -- | --
1K | 10.3 | 0.9 | 2.8 | 0.5
2K | 18.5 | 1.3 | 4.6 | 0.6
4K | 40.5 | 2.2 | 9.6 | 1.1

</body>

</html>
](url)

### CODE REVIEW

**Comment by xuguangxin** [comment]
@bwang30 could you add some data like https://github.com/intel-media-ci/ffmpeg/commit/d4cd8830bdac3e26c8e75cd92e574c159fecc4f7
image

**Comment by bwang30** [comment]
@xuguangxin Thanks for the suggestion, I went through the checkasm files and found that those unit tests seem to be used for performance comparison between original assembly code and optimized assembly code. But for the Sobel filter(which is in vf_convolution.c) in ffmpeg, the original code is a pure c code implementation. Meantime I didn't find any other algorithms such as "boxfilter 3x3, roberts, scharr, prewitt" in the `same file vf_convolution.c` have unit tests code in checkasm folder. Is it suitable to write sobel unit tests into checkasm folder?

**Comment by bwang30** [comment]
@xuguangxin  Add unit test into checkasm and performance data

**Comment by xhaihao** [comment]
there is a typo in commit log
s/optimazation/optimization

**Comment by xhaihao** [comment]
please provide your data in commit log too.

File: libavfilter/x86/vf_convolution_init.c [hunk 1]
```diff
@@ -44,3 +49,17 @@ av_cold void ff_convolution_init_x86(ConvolutionContext *s)
     }
 #endif
 }
+
+av_cold void ff_sobel_init_x86(ConvolutionContext *s)
+{
+#if ARCH_X86_64
+    int i;
+    int cpu_flags = av_get_cpu_flags();
+    for (i = 0; i < 4; i++) {
```
Comment (xhaihao) [inline libavfilter/x86/vf_convolution_init.c]: does s always have 4 planes ? You'd be better to change the condition to 'i < s->nb_planes'

Comment (xhaihao) [reply libavfilter/x86/vf_convolution_init.c]: Community prefers to declare loop indices in the loop statement itself

for (int i = 0; ... )

**Comment by xhaihao** [comment]
> please provide your data in commit log too.

And you could add a description about your optimization in the log.

File: libavfilter/x86/vf_convolution_init.c (hunk 1 above)
Comment (bwang30) [reply libavfilter/x86/vf_convolution_init.c]: Solved in the new commit:
~~~
for (int i = 0; i < s->nb_planes; i++) {
~~~

**Comment by bwang30** [comment]
@xhaihao Merged two commits into one and added performance data from the unit test

**Review by xhaihao — DISMISSED** [review]
LGTM

**Comment by xhaihao** [comment]
https://patchwork.ffmpeg.org/project/ffmpeg/patch/BN6PR11MB1746B88D2FEC63B85650B4379B6D9@BN6PR11MB1746.namprd11.prod.outlook.com/

**Pull request closed without merging**